commit ac3b7425db298aa88f22a4c5a6d0a4c26f7ed338
Author: Thorsten Blum <thorsten.blum@linux.dev>
Date:   Thu Feb 13 12:40:48 2025 +0100

    scsi: hpsa: Replace deprecated strncpy() with strscpy_pad()
    
    strncpy() is deprecated for NUL-terminated destination buffers. Replace
    memset() and strncpy() with strscpy_pad() to copy the version string and
    fill the remaining bytes in the destination buffer with NUL bytes. This
    avoids zeroing the memory before copying the string.
    
    Compile-tested only.
    
    Link: https://github.com/KSPP/linux/issues/90
    Cc: linux-hardening@vger.kernel.org
    Signed-off-by: Thorsten Blum <thorsten.blum@linux.dev>
    Link: https://lore.kernel.org/r/20250213114047.2366-2-thorsten.blum@linux.dev
    Reviewed-by: Bart Van Assche <bvanassche@acm.org>
    Reviewed-by: Kees Cook <kees@kernel.org>
    Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>

diff --git a/drivers/scsi/hpsa.c b/drivers/scsi/hpsa.c
index bb65954b1b1e..c73a71ac3c29 100644
--- a/drivers/scsi/hpsa.c
+++ b/drivers/scsi/hpsa.c
@@ -7230,8 +7230,7 @@ static int hpsa_controller_hard_reset(struct pci_dev *pdev,
 
 static void init_driver_version(char *driver_version, int len)
 {
-	memset(driver_version, 0, len);
-	strncpy(driver_version, HPSA " " HPSA_DRIVER_VERSION, len - 1);
+	strscpy_pad(driver_version, HPSA " " HPSA_DRIVER_VERSION, len);
 }
 
 static int write_driver_ver_to_cfgtable(struct CfgTable __iomem *cfgtable)